openage/opus/opusenc.pyx

```diff
@@ -0,0 +1,415 @@
+# Copyright 2016-2017 the openage authors. See copying.md for legal info.
+
+#!python
+#cython: language_level=3
+#^- asdf: remove if unnecessary?
+
+import time
+from libc.string cimport memcpy, memset
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
+from cpython.bytes cimport PyBytes_FromStringAndSize
+
+from ogg cimport *
+from opus cimport *
+
+verbose = False
+def encode(inputdata=None, be_verbose=False):
+    '''
+    Converts the wav file in the bytes object 'inputdata' to an opusfile
+    and returns it as bytes object. If something fails, a number or an
+    error message is returned.
+    '''
+    global verbose
+    verbose = be_verbose
+
+    inopt = read_wav(inputdata)
+    if type(inopt) != dict:
+        return inopt
+
+    debug("before truncating. len(inputdata) = {}".format(len(inputdata)))
+    debug("trailer_len = {}",inopt['trailer_len'])
+    if inopt['trailer_len']:
+        trailer = inputdata[-inopt['trailer_len']:]
+        inputdata = inputdata[inopt['header_len']:-inopt['trailer_len']]
+    else:
+        inputdata = inputdata[inopt['header_len']:]
+    debug("len(inputdata) = {}".format(len(inputdata)))
+
+    outdata = bytearray(1024*50)
+    cdef size_t outsize = 0 # valid bytes inside 'outdata'
+
+    cdef ogg_packet op
+    cdef ogg_page og
+    cdef ogg_stream_state os
+    cdef OpusEncoder *oe
+    cdef int err
+    cdef int channels = inopt['channels']
+    cdef int wframe_len = inopt['wframe_len']
+    cdef opus_int32 rate = inopt['input_rate']
+    cdef opus_int32 coding_rate = 48000
+    cdef opus_int32 frame_sz = 960
+    cdef opus_int32 lookahead
+
+    if rate > 24000:
+        coding_rate = 48000
+    elif rate > 16000:
+        coding_rate = 24000
+    elif rate > 12000:
+        coding_rate = 16000
+    elif rate > 8000:
+        coding_rate = 12000
+    else:
+        coding_rate = 8000
```

Comment:
I don't understand what your question is. Are you asking, why I didn't choose e.g. coding_rate = 12000 for 12000 <= rate  < 16000 ?
That could lead to aliasing, which sounds horrible, if I didn't lowpass the input below half the coding_rate before resampling. By upsampling we avoid that filtering.